### PR TITLE
[SPARK-10023] [ML] [PySpark] Unified DecisionTreeParams checkpointInterval between Scala and Python API.

### PR DESCRIPTION
**Pull request by yanboliang**
"checkpointInterval" is member of DecisionTreeParams in Scala API which is inconsistency with Python API, we should unified them.

```
member of DecisionTreeParams <-> Scala API
shared param for all ML Transformer/Estimator <-> Python API
```

Proposal:
"checkpointInterval" is also used by ALS, so we make it shared params at Scala.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala [hunk 1]
```diff
@@ -96,23 +96,8 @@ private[ml] trait DecisionTreeParams extends PredictorParams {
     " algorithm will cache node IDs for each instance. Caching can speed up training of deeper" +
     " trees.")
 
-  /**
```
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala]: This scaladoc may still be useful, can we keep it around (maybe as the scaladoc for `setCheckpointInterval`)?

**Comment by feynmanliang** [comment]
To enable checkpointing you also need to set the `cacheNodeIds` param so that should be modified as well.

I think we need to have some more discussion about whether this change is appropriate. I think the rationale for not including in `sharedParams` was to hide the `expertParam`s. It's also important to note that the semantics behind decision tree's `setCheckpointInterval` differs from the current usage of `HasCheckpointInterval` in `ALS` since it depends on `cacheNodeIds`.

CC @jkbradley

File: mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala [hunk 2]
```diff
@@ -56,7 +56,8 @@ private[shared] object SharedParamsCodeGen {
       ParamDesc[String]("inputCol", "input column name"),
       ParamDesc[Array[String]]("inputCols", "input column names"),
       ParamDesc[String]("outputCol", "output column name", Some("uid + \"__output\"")),
-      ParamDesc[Int]("checkpointInterval", "checkpoint interval (>= 1)",
+      ParamDesc[Int]("checkpointInterval", "checkpoint interval (>= 1). E.g. 10 means that " +
+        "the cache will get checkpointed every 10 iterations.", Some("10"),
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala]: Do not set default here. It is hard to make a reasonable default for all subclasses.

File: mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala (hunk 1 above)
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala]: Great idea!

File: mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala [hunk 3]
```diff
@@ -56,7 +56,8 @@ private[shared] object SharedParamsCodeGen {
       ParamDesc[String]("inputCol", "input column name"),
       ParamDesc[Array[String]]("inputCols", "input column names"),
       ParamDesc[String]("outputCol", "output column name", Some("uid + \"__output\"")),
-      ParamDesc[Int]("checkpointInterval", "checkpoint interval (>= 1)",
+      ParamDesc[Int]("checkpointInterval", "checkpoint interval (>= 1). E.g. 10 means that " +
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala]: Maybe it is useful to mention how to disable checkpointing (by setting a large value?).

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks!

File: mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala (hunk 3 above)
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala]: Sorry, I should have seen this PR earlier.  I agree we should make it easy to disable checkpointing.  I was also thinking we could say setting it to -1 disables checkpointing.  (That's clearer than choosing a big number.)

**Comment by jkbradley** [comment]
> I think we need to have some more discussion about whether this change is appropriate. I think the rationale for not including in sharedParams was to hide the expertParams. It's also important to note that the semantics behind decision tree's setCheckpointInterval differs from the current usage of HasCheckpointInterval in ALS since it depends on cacheNodeIds.

@fliang  I agree with this sentiment.  Maybe the best place to put this comment is in the cacheNodeIds param doc.  @yanboliang would you mind sending a little follow-up patch for this JIRA which adds a comment to the cacheNodeIds doc?

**Comment by yanboliang** [comment]
@jkbradley OK, I will send a follow-up PR to add some comments to the cacheNodeIds doc ASAP.

**Pull request closed without merging**